packages/lite-insight/src/attribution/dimensionDrillDownAttribution.ts

```diff
@@ -0,0 +1,89 @@
+import { locatedInInterval, enumerateAllDimensionCombinationsByDFS } from './util';
+import type { Datum } from '../interface';
+import type {
+  DataConfig,
+  CompareInterval,
+  InfoType,
+  FluctInfo,
+  AttributionResult,
+  TreeDim,
+  DataLocation,
+} from './types';
+
+/** Main function for dimension drill down attribution */
+const attribute2MultiDimensions: (
+  sourceData: Datum[],
+  dimensions: string[],
+  measure: string,
+  fluctuationDim: string,
+  baseInterval: CompareInterval,
+  currInterval: CompareInterval
+) => AttributionResult = (sourceData, dimensions, measure, fluctuationDim, baseInterval, currInterval) => {
+  /** remove invalid data */
+  const data = sourceData.filter((item) => !Object.values(item).some((v) => v === null || v === undefined));
+
+  const unitedInfo: InfoType = {
+    baseValue: 0,
+    currValue: 0,
+    diff: 0,
+  };
+
+  const resultTree: TreeDim = {};
+  const DictFlatten = {};
+  /** traverse the input data and build the result data structure; */
+  data.forEach((item) => {
+    let location: DataLocation = 'none';
+    if (locatedInInterval(item[fluctuationDim], baseInterval.startPoint, baseInterval.endPoint)) {
+      location = 'left';
+      unitedInfo.baseValue += item[measure] as number;
+    }
+    if (locatedInInterval(item[fluctuationDim], currInterval.startPoint, currInterval.endPoint)) {
+      location = 'right';
+      unitedInfo.currValue += item[measure] as number;
+    }
+    if (location !== 'none') {
+      const deque: string[] = [];
+      enumerateAllDimensionCombinationsByDFS(
+        item,
+        0,
+        dimensions,
+        resultTree,
+        DictFlatten,
+        deque,
+        measure,
+        fluctuationDim,
+        location
+      );
+    }
+  });
+  unitedInfo.diff = unitedInfo.currValue - unitedInfo.baseValue;
+
+  return { resultTree, unitedInfo, resultFlatten: Object.values(DictFlatten) };
+};
+
+export class DimensionDrillDownAttribution {
```

Comment:
Why use class instead of function？

If function return `{wholeData, treeData, flattenData}`，it works well.

Reply:
Different users require different return types.  If all the data are returned in a single result, this result could be redundant. So the attribution class is abstracted, which is capable of returning different data formats regarding to different requirements.

Reply:
Reserved for functional scalability, OK